File: backend/routers/release_rollouts.py
"""Headquarters-only staged release control plane; it records decisions but never deploys."""

from __future__ import annotations

from typing import Literal

from core.database import get_db
from dependencies.auth import get_current_user
from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel, Field
from schemas.auth import UserResponse
from services.audit import record_audit_event
from services.release_rollouts import ReleaseRolloutService
from services.tenant_access import require_global_platform_access
from sqlalchemy.ext.asyncio import AsyncSession


router = APIRouter(prefix="/api/v1/release-rollouts", tags=["release-rollouts"])


class RolloutCreate(BaseModel):
    version: str = Field(min_length=1, max_length=100)
    release_role: Literal["hq", "agency", "client"]
    deployment_id: str = Field(min_length=1, max_length=100)
    manifest_sha256: str = Field(pattern=r"^[a-fA-F0-9]{64}$")
    change_summary: str | None = Field(default=None, max_length=2000)


class RolloutAction(BaseModel):
    stage_key: Literal["hq", "test_agency", "test_client_plan", "full_rollout"]
    action: Literal["start", "approve", "fail"]
    note: str | None = Field(default=None, max_length=2000)


class RollbackRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=1000)


async def _audit(db: AsyncSession, request: Request, user: UserResponse, action: str, rollout_id: int, detail: dict) -> None:
    record_audit_event(db, action=action, actor_user_id=user.id, target_type="release_rollout", target_id=rollout_id, ip_address=request.client.host if request.client else None, detail=detail)
    await db.commit()


@router.get("")
async def list_rollouts(db: AsyncSession = Depends(get_db), current_user: UserResponse = Depends(get_current_user)):
    await require_global_platform_access(current_user=current_user)
    return {"items": await ReleaseRolloutService(db).list()}


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_rollout(payload: RolloutCreate, request: Request, db: AsyncSession = Depends(get_db), current_user: UserResponse = Depends(get_current_user)):
    await require_global_platform_access(current_user=current_user)
    try:
        result = await ReleaseRolloutService(db).create({**payload.model_dump(), "created_by": current_user.id})
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    await _audit(db, request, current_user, "release_rollout_created", result["id"], {"version": result["version"], "deployment_id": result["deployment_id"]})
    return result


@router.post("/{rollout_id}/action")
async def act_rollout(rollout_id: int, payload: RolloutAction, request: Request, db: AsyncSession = Depends(get_db), current_user: UserResponse = Depends(get_current_user)):
    await require_global_platform_access(current_user=current_user)
    try:
        result = await ReleaseRolloutService(db).act(rollout_id, actor=current_user.id, **payload.model_dump())
    except KeyError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    await _audit(db, request, current_user, "release_rollout_stage_action", rollout_id, {"stage": payload.stage_key, "action": payload.action})
    return result


@router.post("/{rollout_id}/rollback")
async def rollback_rollout(rollout_id: int, payload: RollbackRequest, request: Request, db: AsyncSession = Depends(get_db), current_user: UserResponse = Depends(get_current_user)):
    await require_global_platform_access(current_user=current_user)
    try:
        result = await ReleaseRolloutService(db).rollback(rollout_id, reason=payload.reason, actor=current_user.id)
    except KeyError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    await _audit(db, request, current_user, "release_rollout_rolled_back", rollout_id, {"reason": payload.reason})
    return result

File: frontend/src/lib/release-rollouts.ts
import { authApi } from "@/lib/auth";
import { getAPIBaseURL } from "@/lib/config";

const BASE = "/api/v1/release-rollouts";

async function request<T>(path: string, init: RequestInit = {}): Promise<T> {
  const headers = new Headers(init.headers);
  headers.set("Content-Type", "application/json");
  const token = authApi.getStoredToken();
  if (token) headers.set("Authorization", `Bearer ${token}`);
  const response = await fetch(`${getAPIBaseURL()}${BASE}${path}`, { ...init, headers });
  if (!response.ok) throw new Error(`发布控制请求失败：${response.status}`);
  return response.json() as Promise<T>;
}

export type ReleaseStage = { stage_key: string; stage_label: string; sequence: number; status: string; note?: string | null };
export type ReleaseRollout = { id: number; version: string; release_role: string; deployment_id: string; manifest_sha256: string; change_summary?: string | null; status: string; current_stage?: string | null; rollback_reason?: string | null; stages: ReleaseStage[] };

export const releaseRolloutsApi = {
  list: async () => (await request<{ items: ReleaseRollout[] }>("")).items,
  create: (payload: { version: string; releaseRole: "hq" | "agency" | "client"; deploymentId: string; manifestSha256: string; changeSummary?: string }) => request<ReleaseRollout>("", { method: "POST", body: JSON.stringify({ version: payload.version, release_role: payload.releaseRole, deployment_id: payload.deploymentId, manifest_sha256: payload.manifestSha256, change_summary: payload.changeSummary || null }) }),
  action: (rolloutId: number, stageKey: string, action: "start" | "approve" | "fail", note?: string) => request<ReleaseRollout>(`/${rolloutId}/action`, { method: "POST", body: JSON.stringify({ stage_key: stageKey, action, note: note || null }) }),
  rollback: (rolloutId: number, reason: string) => request<ReleaseRollout>(`/${rolloutId}/rollback`, { method: "POST", body: JSON.stringify({ reason }) }),
};
